### PR TITLE
[TKW] Masking

### PR DESCRIPTION
**Pull request by Hardcode84**
Add support for cases when input sizes are not divisible by WG sizes, use masked vector instructions int his case, to only access elements within arrays bounds.

### CODE REVIEW

**Comment by Hardcode84** [comment]
e2e tests now all pass locally with weird sizes, so it's ready for review. Bu I'll still wait until e2e fixed in main before merging.

File: tests/kernel/wave/wave_e2e_test.py
Comment (harsh-nod) [inline tests/kernel/wave/wave_e2e_test.py:560]: Not for this patch, but would it work if you parameterized h, w, hf and wf?

File: tests/kernel/wave/wave_e2e_test.py
Comment (harsh-nod) [inline tests/kernel/wave/wave_e2e_test.py:16]: nice!

File: tests/kernel/wave/wave_e2e_test.py
Comment (harsh-nod) [inline tests/kernel/wave/wave_e2e_test.py:54]: What happens if you remove the max and leave it to min(N, 256)? Just wondering why the max is necessary.

File: shark_turbine/kernel/wave/constraints.py
Comment (harsh-nod) [inline shark_turbine/kernel/wave/constraints.py:174]: since this is the total number of workgroups, should we just call it that? num_workgroups() ? (iterations is a little misleading since usually it is used in the context of a for-loop which in our case would be the tkw.reduction / TilingConstraint)

File: shark_turbine/kernel/wave/codegen.py
Comment (harsh-nod) [inline shark_turbine/kernel/wave/codegen.py:920]: Not for this PR, but wondering whether we should have subs_idxc return an int when we know the return type can be converted to an int? This way subs_idxc and safe_subs are guaranteed to give either floats/ints or symbolic expressions.

File: shark_turbine/kernel/wave/codegen.py
Comment (harsh-nod) [inline shark_turbine/kernel/wave/codegen.py:386]: Can you add a lit test showing what the IR looks like?

File: shark_turbine/kernel/wave/codegen.py
Comment (harsh-nod) [inline shark_turbine/kernel/wave/codegen.py:402]: Doesn't get_sympy_index(dim) = dim ? So why do we need this?

File: shark_turbine/kernel/wave/codegen.py
Comment (harsh-nod) [reply shark_turbine/kernel/wave/codegen.py:386]: Also helps for me to debug and step through locally without having access to the cluster.

**Review by harsh-nod — CHANGES_REQUESTED** [review]
Could you add a lit test that shows the IR? I will be able to follow better if I can step through the code and understand what's happening. Thanks!

File: tests/kernel/wave/wave_e2e_test.py
Comment (Hardcode84) [reply tests/kernel/wave/wave_e2e_test.py:685]: As I'm testing 3*4*3*3 total permutations now, it's infeasible to find good seed for each one and I'm getting occasional mismatches on `1e-05`.

File: tests/kernel/wave/wave_e2e_test.py
Comment (Hardcode84) [reply tests/kernel/wave/wave_e2e_test.py:560]: We should, but I need to find a better way of deciding which combination of parameters we are going to test, as currently amount of test permutations grow multiplicative.

File: shark_turbine/kernel/wave/constraints.py
Comment (Hardcode84) [reply shark_turbine/kernel/wave/constraints.py:174]: I choose `iterations` specifically, because it's used interchangeably between `WorkgroupConstraint` and `TilingConstraint` in bounds check so I don't need to check for them separately. Maybe any ideas for better common name?

File: tests/kernel/wave/wave_e2e_test.py
Comment (Hardcode84) [reply tests/kernel/wave/wave_e2e_test.py:54]: I want to keep block size no smaller than `wave_size`, but I should probably change it  `sympy.Max(sympy.Min(N, 256), wave_size)` to make it clear

Comment (Hardcode84) [reply tests/kernel/wave/wave_e2e_test.py]: updated

File: shark_turbine/kernel/wave/codegen.py
Comment (Hardcode84) [reply shark_turbine/kernel/wave/codegen.py:386]: done

File: tests/kernel/wave/wave_e2e_test.py
Comment (harsh-nod) [reply tests/kernel/wave/wave_e2e_test.py:685]: Okay makes sense. Thanks!

File: tests/kernel/wave/wave_e2e_test.py
Comment (harsh-nod) [reply tests/kernel/wave/wave_e2e_test.py:560]: Makes sense. Just wanted to know if it was working and sounds like it is which is great. We can parameterize it for the nightly test.

File: tests/kernel/wave/wave_e2e_test.py
Comment (harsh-nod) [reply tests/kernel/wave/wave_e2e_test.py]: thanks

File: shark_turbine/kernel/wave/constraints.py
Comment (harsh-nod) [reply shark_turbine/kernel/wave/constraints.py:174]: How about counts? Specifies number of loop counts or workgroup counts?

File: shark_turbine/kernel/wave/codegen.py
Comment (harsh-nod) [reply shark_turbine/kernel/wave/codegen.py:386]: thanks!

File: shark_turbine/kernel/wave/codegen.py
Comment (harsh-nod) [reply shark_turbine/kernel/wave/codegen.py:402]: that explains it, yeah agree a better name would help.

File: lit_tests/kernel/wave/codegen.py
Comment (harsh-nod) [inline lit_tests/kernel/wave/codegen.py:256]: This is exactly what I was looking for - thank you. So this brings me to my main question, why do we need to do all this arithmetic to compute the mask? Since we have the information from sympy in the python layer, can we figure out what the mask should be before codegen? The advantage of that is we avoid all these extra instructions.

Comment (harsh-nod) [reply lit_tests/kernel/wave/codegen.py:256]: If the shape was dynamic, then I think we would have to do this, but if it was static I think we could avoid this?

**Review by harsh-nod — APPROVED** [review]
lgtm, modulo changes we discussed!

**Pull request merged**
